File: estagio/conting/conting/modif.c
/*------------------------------modif.c-----------------------------*/

/* Funcao para guardar/reestabelecer Yo do caso base  */

#include <stdio.h>
#include <stdlib.h>
#include "global.h"
/*-----------------------Inicio da funcao ricupero--------------------*/
void ricupero(int info){
	int i; 
	/*---------------------------------------------------------------------
	Guarda a matriz Yo=Go + j.Bo do caso base (para posterior utilizacao)
	LEGENDA:  info = 0   ==> Guarda  a matriz Yo = Go + j.Bo  do caso base p/ posterior utiliz.
	info = 1   ==> Reestabelece a matriz Yo = Go + j.Bo a sua forma original
	-----------------------------------------------------------------------*/
	if(info == 0){	
	/*-------------------------------------------------------
	Guarda o estado do caso base para posterior utilizacao
	------------------------------------------------------*/
		for(i=0; i<nb; i++){  // Diagonal
			gdiagbase[i] = gdiag[i];
			bdiagbase[i] = bdiag[i];     
		}
		for(i=0; i<nl; i++){       // Elementos do ramo
			gklbase[i] = gkl[i];
			bklbase[i] = bkl[i];
		//        printf("\n GUARDANDO...bklbase[%d]=%lf \n",li,bklbase[li]);
		//        printf("\n GUARDANDO...bkl[%d]=%lf \n",li,bkl[li]);
		}
		for(i=0; i<nb ;i++){
			vb[i] = v[i];
			tetab[i] = tet[i];
		}
	}  // fim if
	if(info == 1){
	/*-------------------------------------------------
	Guarda a matriz Yo=Go + j.Bo a sua forma original
	------------------------------------------------*/
		for(i=0; i<nb; i++){   // Diagonal
			gdiag[i] = gdiagbase[i];
			bdiag[i] = bdiagbase[i];
		}
		for(i=0; i<nl; i++){          // Elementos do ramo
			gkl[i] = gklbase[i];
			bkl[i] = bklbase[i];
	//printf("\n RECUPERANDO...bklbase[%d]=%lf \n",li,bklbase[li]);
	//printf("\n RECUPERANDO...bkl[%d]=%lf \n",li,bkl[li]);  
		}
	/*-Restabelece o estado do caso base para utilizacao no proximo caso-----*/
		for(i=0; i<nb ;i++){
			v[i] = vb[i];
			tet[i] = tetab[i];
		}
	}    // fim if
	return;
}
/*--------------------------Fim da funcao ricupero---------------------*/
/*-------------------------Inicio da funcao modify---------------------*/
/* Funcao para modificar a matriz Y em funcao da contingencia
OBS: fcont e o fator de contingencias de ramo [-1.<= fcont =<+1.]
    fcont e introduzido para permitir a retirada de uma fracao da
    admitancia do ramo. Se fcont-1.0 tira-se o ramo integralmente. 

OBS: ramos paralelos (transformador ou linha) sao armazenados
     individualmente nos vetores gdiag[],bdiag[] e bkl[]

OBS: ter bastante cuidado com a maneira como estao armazenados os 
     parametros do circuito PI de transformadores em fase
*/

void modify(int j,int k,int i){ 
	double t;
	//printf("fcont  = %f", fcont);
	if(tap[i] == 0.e0) {          // Linha de transmissao
		gdiag[j] = gdiag[j] - fcont * gkl[i];
		gdiag[k] = gdiag[k] - fcont * gkl[i];
		bdiag[j] = bdiag[j] - fcont * bkl[i] - fcont * ylin[i];
		bdiag[k] = bdiag[k] - fcont * bkl[i] - fcont * ylin[i];
	}
	else{
		t = 1.e0/tap[i];
		gdiag[j] = gdiag[j] - fcont * gkl[i] * t;
		gdiag[k] = gdiag[k] - fcont * gkl[i] / t;
		bdiag[j] = bdiag[j] - fcont * bkl[i] * t;
		bdiag[k] = bdiag[k] - fcont * bkl[i] / t;
	}
	gkl[i] = gkl[i] - fcont * gkl[i];
	bkl[i] = bkl[i] - fcont * bkl[i];
	return;
}

/*-----------------------------Fim da funcao modify------------------------*/

File: estagio/conting/PIPELINE/modif.f
c23456781234567812345678123456781234567812345678123456781234567812345678
c   1       2       3       4       5       6       7       8       9
c ----------------------------------------------------------------------
c
c ----------------------------------------------------------------------
c Subrotina para guardar/restabelecer Yo e estado do caso base 
     
      subroutine ricupero(info)

      include 'globvar.inc'
      include 'varcont.inc'
c
      character*8 info

  
      if( info .eq. 'guarda' ) then

c Guarda a matriz Yo = Go + j.Bo do caso base (para posterior utilizacao)

      do 6 kbus = 1,nb                 ! Diagonal
        gdiagbase(kbus) = gdiag(kbus)   
        bdiagbase(kbus) = bdiag(kbus) 
    6 continue
     
      do 7 li = 1,nl                   ! Elementos de ramo 
        gklbase(li) = gkl(li)          
        bklbase(li) = bkl(li)
       
cc        print*,'GUARDANDO .... bklbase(',li,') =',bklbase(li)
cc        print*,'GUARDANDO .... bkl(',li,') =',bkl(li)
         
    7 continue


c Guarda o estado do caso base <para posterior utilizacao>

      do iii = 1,nb   ! guarda o caso base em vb() e tetab()
        vb(iii) = v(iii)
        tetab(iii) = tet(iii)
      enddo
     
      endif
      
      if( info .eq. 'recupera' ) then

c Restabelece a matriz Yo = Go + j.Bo a sua forma original

      do 60 kbus = 1,nb                 ! Diagonal
        gdiag(kbus) = gdiagbase(kbus)   
        bdiag(kbus) = bdiagbase(kbus)
   60 continue
     
      do 70 li = 1,nl                   ! Elementos de ramo
        gkl(li) = gklbase(li)         
        bkl(li) = bklbase(li)
        
cc        print*,'RECUPERANDO .... bklbase(',li,') =',bklbase(li)
cc        print*,'RECUPERANDO .... bkl(',li,') =',bkl(li)
       
   70 continue


c Restabelece o estado do caso base <para utilizacao no proximo caso>

      do iii = 1,nb         
        v(iii) = vb(iii)
        tet(iii) = tetab(iii)
      enddo

      endif

      return
      end         ! Fim de ricupero

c
c ----------------------------------------------------------------------
c Subrotina para modificar a matriz Y em funcao da contingencia

      subroutine modify(j1,k1,ii) 
c
      include 'globvar.inc'
      include 'varcont.inc' 

c OBS:  fcont e o fator de contingencias de ramo [-1.<= fcont =<+1.] <var. global>
c       fcont e introduzido para permitir a retirada de uma fracao da admitancia 
c       do ramo. Se fcont = 1.0 tira-se o ramo integralmente.

c OBS: ramos paralelos (transformador ou linha) sao armazenados individualmente
c      nos vetores gdiag(), bdiag(), gkl() e bkl()

c OBS: ter bastante cuidado com a maneira como estao armazenados os parametros do
c      circuito PI de transformadores em-fase

     
      if( tap(ii) .eq. 0.0 ) then            ! linha de transmissao
          gdiag(j1) = gdiag(j1) - fcont*gkl(ii)
          gdiag(k1) = gdiag(k1) - fcont*gkl(ii)
          bdiag(j1) = bdiag(j1) - fcont*bkl(ii) - fcont*ylin(ii)
          bdiag(k1) = bdiag(k1) - fcont*bkl(ii) - fcont*ylin(ii)
      else                                   ! transformador
          t1 = 1./tap(ii)
          gdiag(j1) = gdiag(j1) - fcont*gkl(ii)*t1
          gdiag(k1) = gdiag(k1) - fcont*gkl(ii)/t1
          bdiag(j1) = bdiag(j1) - fcont*bkl(ii)*t1
          bdiag(k1) = bdiag(k1) - fcont*bkl(ii)/t1                    
      endif

      gkl(ii) = gkl(ii) - fcont*gkl(ii)       
      bkl(ii) = bkl(ii) - fcont*bkl(ii)                     

      return
      end          ! Fim da subrotina modify
